velox/exec/tests/SparkShuffleTest.cpp

```diff
@@ -0,0 +1,611 @@
+    });
+  }
+
+  RowTypePtr makeRowType(const std::vector<VectorPtr>& args) const {
+    std::vector<std::string> inputNames;
+    std::vector<TypePtr> inputTypes;
+    inputNames.reserve(args.size());
+    inputTypes.reserve(args.size());
+
+    for (auto i = 0; i < args.size(); ++i) {
+      inputNames.push_back(fmt::format("c{}", i));
+      inputTypes.push_back(args[i]->type());
+    }
+
+    return ROW(std::move(inputNames), std::move(inputTypes));
+  }
+
+  static std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+    return {exec::FunctionSignatureBuilder()
+                .argumentType("any")
+                .variableArity()
+                .returnType("varbinary")
+                .build()};
+  }
+};
+
+class TestShuffle : public exec::Shuffle {
+ public:
+  TestShuffle(
+      memory::MemoryPool* pool,
+      uint32_t numPartitions,
+      uint32_t maxBytesPerPartition)
+      : pool_{pool},
+        numPartitions_{numPartitions},
+        maxBytesPerPartition_{maxBytesPerPartition},
+        inProgressSizes_(numPartitions, 0) {
+    inProgressPartitions_.resize(numPartitions_);
+    readyPartitions_.resize(numPartitions_);
+  }
+
+  void collect(int32_t partition, std::string_view data) {
+    auto& buffer = inProgressPartitions_[partition];
+
+    // Check if there is enough space in the buffer.
+    if (buffer &&
+        inProgressSizes_[partition] + data.size() + sizeof(size_t) >=
+            maxBytesPerPartition_) {
+      buffer->setSize(inProgressSizes_[partition]);
+      readyPartitions_[partition].emplace_back(std::move(buffer));
+      inProgressPartitions_[partition].reset();
+    }
+
+    // Allocate buffer if needed.
+    if (!buffer) {
+      buffer = AlignedBuffer::allocate<char>(maxBytesPerPartition_, pool_);
+      inProgressSizes_[partition] = 0;
+    }
+
+    // Copy data.
+    auto rawBuffer = buffer->asMutable<char>();
+    auto offset = inProgressSizes_[partition];
+
+    *(size_t*)(rawBuffer + offset) = data.size();
+
+    offset += sizeof(size_t);
+    memcpy(rawBuffer + offset, data.data(), data.size());
+
+    inProgressSizes_[partition] += sizeof(size_t) + data.size();
+  }
+
+  void noMoreData() {
+    for (auto i = 0; i < numPartitions_; ++i) {
+      if (inProgressSizes_[i] > 0) {
+        auto& buffer = inProgressPartitions_[i];
+        buffer->setSize(inProgressSizes_[i]);
+        readyPartitions_[i].emplace_back(std::move(buffer));
+        inProgressPartitions_[i].reset();
+      }
+    }
+  }
+
+  bool hasNext(int32_t partition) const {
+    return !readyPartitions_[partition].empty();
+  }
+
+  BufferPtr next(int32_t partition) {
+    VELOX_CHECK(!readyPartitions_[partition].empty());
+
+    auto buffer = readyPartitions_[partition].back();
+    readyPartitions_[partition].pop_back();
+    return buffer;
+  }
+
+ private:
+  memory::MemoryPool* pool_;
+  const uint32_t numPartitions_;
+  const uint32_t maxBytesPerPartition_;
+  std::vector<BufferPtr> inProgressPartitions_;
```

Comment:
At first, I was a little confused why despite the fact that buffers have sizes you used another array for sizes but I think I see the ease with using it but would using a adjusting buffer sizes in this case practical as well?

Reply:
I think that can work as well.